File: internal/controller/scriptdefinition/execution.go
package scriptdefinition

import (
	requestidempotency "github.com/AlanD20/groundplane/internal/controller/idempotency"
	"github.com/AlanD20/groundplane/internal/core"
	apiTypes "github.com/AlanD20/groundplane/pkg/api"
)

func executionFromAPI(input *apiTypes.ScriptExecution) *core.ScriptExecution {
	if input == nil || (input.Mode == "inherited" && input.Image == "" && input.User == "" &&
		input.Volumes == nil && input.EntryIDs == nil) {
		return nil
	}
	execution := &core.ScriptExecution{Mode: core.ScriptExecutionMode(input.Mode), Image: input.Image, User: input.User,
		EntryIDs: input.EntryIDs}
	if input.Volumes != nil {
		execution.Volumes = make([]core.ScriptVolumeGrant, 0, len(input.Volumes))
		for _, grant := range input.Volumes {
			execution.Volumes = append(execution.Volumes, core.ScriptVolumeGrant{
				VolumeID: grant.VolumeID, Target: grant.Target, ReadOnly: grant.ReadOnly,
			})
		}
	}
	return execution
}

func executionResponse(execution *core.ScriptExecution) apiTypes.ScriptExecution {
	if execution == nil {
		return apiTypes.ScriptExecution{Mode: "inherited"}
	}
	response := apiTypes.ScriptExecution{Mode: string(execution.Mode), Image: execution.Image, User: execution.User,
		EntryIDs: execution.EntryIDs}
	for _, grant := range execution.Volumes {
		response.Volumes = append(response.Volumes, apiTypes.ScriptVolumeGrant{
			VolumeID: grant.VolumeID, Target: grant.Target, ReadOnly: grant.ReadOnly,
		})
	}
	return response
}

func executionIntent(execution apiTypes.ScriptExecution) requestidempotency.Value {
	fields := []requestidempotency.Field{{Name: "mode", Value: requestidempotency.String(execution.Mode)}}
	if execution.Mode == "inherited" {
		return requestidempotency.Object(fields...)
	}
	volumes := make([]requestidempotency.Value, 0, len(execution.Volumes))
	for _, grant := range execution.Volumes {
		volumes = append(volumes, requestidempotency.Object(
			requestidempotency.Field{Name: "volume_id", Value: requestidempotency.String(grant.VolumeID)},
			requestidempotency.Field{Name: "target", Value: requestidempotency.String(grant.Target)},
			requestidempotency.Field{Name: "read_only", Value: requestidempotency.Bool(grant.ReadOnly)},
		))
	}
	entries := make([]requestidempotency.Value, 0, len(execution.EntryIDs))
	for _, id := range execution.EntryIDs {
		entries = append(entries, requestidempotency.String(id))
	}
	fields = append(fields,
		requestidempotency.Field{Name: "image", Value: requestidempotency.String(execution.Image)},
		requestidempotency.Field{Name: "user", Value: requestidempotency.String(execution.User)},
		requestidempotency.Field{Name: "volumes", Value: requestidempotency.List(volumes...)},
		requestidempotency.Field{Name: "entry_ids", Value: requestidempotency.List(entries...)},
	)
	return requestidempotency.Object(fields...)
}

File: console/src/features/script/execution.ts
import type { ScriptExecution } from "./types";

export const maximumScriptVolumes = 32;
export const maximumScriptEntries = 64;

const reservedTargets = [
  "/groundplane-script-body",
  "/proc",
  "/sys",
  "/dev",
  "/run",
  "/var/run",
  "/bin",
  "/sbin",
  "/usr",
  "/lib",
  "/lib64",
  "/etc/hosts",
  "/etc/hostname",
  "/etc/resolv.conf",
];

function overlaps(left: string, right: string): boolean {
  return (
    left === right ||
    left.startsWith(`${right}/`) ||
    right.startsWith(`${left}/`)
  );
}

export function scriptExecutionError(
  execution: ScriptExecution,
): string | null {
  if (execution.mode === "inherited") return null;
  // The Controller remains authoritative for the complete OCI name grammar.
  const pinned = /^([^\s@]+)@sha256:[a-f0-9]{64}$/.exec(execution.image);
  const repository = pinned?.[1].split("/").at(-1);
  if (!repository || repository.includes(":"))
    return "Execution image must be a repository@sha256:digest reference without a tag.";
  if (
    !/^(0|[1-9][0-9]*):(0|[1-9][0-9]*)$/.test(execution.user) ||
    execution.user.split(":").some((part) => Number(part) > 4294967295)
  ) {
    return "Execution user must be numeric uid:gid (for example, 0:0).";
  }
  if (
    execution.volumes.length > maximumScriptVolumes ||
    execution.entryIds.length > maximumScriptEntries
  ) {
    return "Execution supports at most 32 Volume grants and 64 Entry grants.";
  }
  const seenVolumes = new Set<string>();
  for (const [index, grant] of execution.volumes.entries()) {
    if (
      !/^vol_[0-9A-HJKMNP-TV-Z]{26}$/.test(grant.volumeId) ||
      seenVolumes.has(grant.volumeId)
    ) {
      return "Execution requires a unique, stable Volume selection for each grant.";
    }
    seenVolumes.add(grant.volumeId);
    if (
      !grant.target.startsWith("/") ||
      grant.target === "/" ||
      /[\u0000-\u001f\u007f-\u009f]/u.test(grant.target) ||
      grant.target
        .slice(1)
        .split("/")
        .some((part) => part === "" || part === "." || part === "..")
    ) {
      return "Execution Volume targets must be canonical absolute paths, not root or traversal paths.";
    }
    if (
      reservedTargets.some((path) => overlaps(path, grant.target)) ||
      execution.volumes
        .slice(0, index)
        .some((previous) => overlaps(previous.target, grant.target))
    ) {
      return "Execution Volume targets must not overlap each other or reserved system paths.";
    }
  }
  if (
    new Set(execution.entryIds).size !== execution.entryIds.length ||
    execution.entryIds.some((id) => !/^ev_[0-9A-HJKMNP-TV-Z]{26}$/.test(id))
  ) {
    return "Execution Entry selections must be unique stable ids.";
  }
  return null;
}
